<a href="https://colab.research.google.com/github/Ssonamy/MADSU/blob/main/Lab1.2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Лабораторная работа №1.2
Задания для самостоятельного выполнения:

## **Импорт библиотек**

In [ ]:
import pandas as pd
import numpy as np
import kagglehub

path = kagglehub.dataset_download("brendan45774/test-file")
# /kaggle/input/test-file/tested.csv
# Путь к файлу
print("Path to dataset files:", path)

Using Colab cache for faster access to the 'test-file' dataset.
Path to dataset files: /kaggle/input/test-file


## **Библиотека NumPy**

### **1. Сгенерируйте одномерный NumPy массив длины 10, состоящий из случайных чисел, взятых из нормального распределения (со средним 0 и среднеквадратическим отклонением 1). Выполните следующие задания:**


*   Выведите последние 3 значения массива
*   Отсортируйте значения по убыванию, выведите индексы двух самых больших значений массива
*   Найдите норму



Для фиксирования воспросизводимости результатов укажем seed для генерации псевдослучайных чисел

In [ ]:
np.random.seed(0)

Генерация массива длинны 10 из случайных чисел, взятых из нормального распределения

In [ ]:
a1 = np.random.randn(10)
a1

array([ 1.76405235,  0.40015721,  0.97873798,  2.2408932 ,  1.86755799,
       -0.97727788,  0.95008842, -0.15135721, -0.10321885,  0.4105985 ])

Выведем 3 последних значения:

In [ ]:
a1[-3:]

array([-0.15135721, -0.10321885,  0.4105985 ])

Отсортируем массив по убыванию:

In [ ]:
np.sort(a1)[::-1]

array([ 2.2408932 ,  1.86755799,  1.76405235,  0.97873798,  0.95008842,
        0.4105985 ,  0.40015721, -0.10321885, -0.15135721, -0.97727788])

Выведем индексы, двух самых больших значений массива:

In [ ]:
np.argsort(a1)[-2::]

array([4, 3])

Найдем норму:

In [ ]:
print(np.linalg.norm(a1))

3.8469733571119606


### **2. Сгенерируйте одномерный NumPy массив, состоящий из чётных чисел с 10 до 40 включительно. Выполните следующие задания:**


*   Переведите его в квадратную матрицу (двумерный NumPy массив). Транспонируйте её.

*   Найдите определитель транспонированной матрицы.



Создадим numpy-массив со значениями от 10 до 40 включительно:
Для этого воспользуемся методом .arange(), создающим массив в заданом диапазоне с заданым шагом

In [ ]:
a2 = np.arange(10, 40 + 1, 2)
a2

array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40])

Переведем полученный массив в квадратную матрицу:

In [ ]:
a2.resize(4, 4)
a2

array([[10, 12, 14, 16],
       [18, 20, 22, 24],
       [26, 28, 30, 32],
       [34, 36, 38, 40]])

Транспонируем матрицу:

In [ ]:
a2.transpose()

array([[10, 18, 26, 34],
       [12, 20, 28, 36],
       [14, 22, 30, 38],
       [16, 24, 32, 40]])

Находим определитель матрицы:

In [ ]:
print(np.linalg.det(a2))

0.0


### **З. Сгенерируйте случайный целочисленный одномерный NumPy массив длинны 15, значения для которого берутся из диапазона от 1 до 30. Найдите его минимальное и максимальное значения. Создайте массив со значениями между найденными минимальным и максимальным значениями с равномерным шагом, содержащий 20 значений.**

Создадим одномерный массив длинны 15:

In [ ]:
a3 = np.random.randint(1, 30, 15)
a3

array([ 7,  8, 24, 15, 25, 18,  6, 26, 14,  9, 10, 21, 20, 17, 20])

Найдем минимальное и максимальное значения:

In [ ]:
print(a3.min())

6


In [ ]:
print(a3.max())

26


Создадим массив со значениями между минимальным и максимальным значениями с равномерным шагом длинны 20:

In [ ]:
a3a = np.linspace(a3.min(), a3.max(), 20)
a3a

array([ 6.        ,  7.05263158,  8.10526316,  9.15789474, 10.21052632,
       11.26315789, 12.31578947, 13.36842105, 14.42105263, 15.47368421,
       16.52631579, 17.57894737, 18.63157895, 19.68421053, 20.73684211,
       21.78947368, 22.84210526, 23.89473684, 24.94736842, 26.        ])

### **4. Сгенерируйте матрицу А, как случайный целочисленный двумерный NumPy массив размера 6 х 5, значения для которого берутся из диапазона от 1 до 10. Выполните следующие задания:**

*   Присоедините к этой матрице слева столбец из единиц (обозначим такую матрицу A˜)
*   Найдите A˜T
*   Найдите произведение матриц A˜T A˜
*   Найдите обратную матрицу  (A˜T A˜)^−1

Создадим матрицу А, как случайный целочисленный двумерный NumPy массив размера б х 5, значения для которого берутся из диапазона от 1 до 10:

In [ ]:
A = np.random.randint(1, 10, (6, 5))
A

array([[6, 1, 3, 4, 9],
       [2, 4, 4, 4, 8],
       [1, 2, 1, 5, 8],
       [4, 3, 8, 3, 1],
       [1, 5, 6, 6, 7],
       [9, 5, 2, 5, 9]])

Создадим и присоединим к этой матрице слева столбец из единиц:

In [ ]:
B = np.ones((6, 1), dtype = "int")
B

array([[1],
       [1],
       [1],
       [1],
       [1],
       [1]])

In [ ]:
A1 = np.column_stack((A, B)) #А штрих
A1

array([[6, 1, 3, 4, 9, 1],
       [2, 4, 4, 4, 8, 1],
       [1, 2, 1, 5, 8, 1],
       [4, 3, 8, 3, 1, 1],
       [1, 5, 6, 6, 7, 1],
       [9, 5, 2, 5, 9, 1]])

Транспонируем матрицу А штрих:

In [ ]:
A2 = A1.T # А штрих транспонированное
A2

array([[6, 2, 1, 4, 1, 9],
       [1, 4, 2, 3, 5, 5],
       [3, 4, 1, 8, 6, 2],
       [4, 4, 5, 3, 6, 5],
       [9, 8, 8, 1, 7, 9],
       [1, 1, 1, 1, 1, 1]])

Найдем произведение А штрих транспонированное и А штрих

In [ ]:
A3 = A1 @ A2 # Произведение матриц
A3

array([[144, 117, 104,  73, 117, 167],
       [117, 117,  99,  73, 127, 139],
       [104,  99,  96,  42, 104, 119],
       [ 73,  73,  42, 100,  93,  92],
       [117, 127, 104,  93, 148, 140],
       [167, 139, 119,  92, 140, 217]])

Найдем обратную матрицу (А штрих транспонированное * А штрих) ^ -1:

In [ ]:
A4 = np.linalg.inv(A3)
A4

array([[ 1.51384083, -1.57608926, -2.33320387, -1.54865476,  2.78161853,
        -0.01398206],
       [-1.57608926,  2.08174434,  2.35199722,  1.64903681, -3.1662593 ,
        -0.06672033],
       [-2.33320387,  2.35199722,  3.88907232,  2.54485588, -4.46772539,
        -0.04021825],
       [-1.54865476,  1.64903681,  2.54485588,  1.70820837, -3.01621441,
        -0.03831232],
       [ 2.78161853, -3.1662593 , -4.46772539, -3.01621441,  5.48809248,
         0.07556612],
       [-0.01398206, -0.06672033, -0.04021825, -0.03831232,  0.07556612,
         0.04765243]])

Проверим, что обратная матрица найдена правильно (так как есть погрешность, идеальная единичная матрица не получается, для этого используем медод .allclose, который выводит True/False, в зависимости от того, близко ли значение матриц):

In [ ]:
A5 = A3 @ A4
print(np.allclose(A5, np.eye(6)))

True


## **Библиотека pandas**

### **5. Загрузите с kaggle.com датасет о пассажирах Титаника. Считайте файл в Pandas DataFrame. Создайте копию и выведите первые 10 строк. Выведите третью, десятую и последнюю строки в одгном датафрейме.**

Указание:
данные можно скачать тут: https://www.kaggle.com/datasets/fedesoriano/the-boston-houseprice-data

*Вероятно, необходимо будет найти правильную ссылку на нужный датасет.*

Создадим копию и выведем первые десять строк:

In [ ]:
data = pd.read_csv("/kaggle/input/test-file/tested.csv")
data.head(10)

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   
5          897         0       3   
6          898         1       3   
7          899         0       2   
8          900         1       3   
9          901         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
5                    Svensson, Mr. Johan Cervin    male  14.0      0      0   
6                          Connolly, Miss. Kate  female  30.0      0      0   
7                  Caldwell, Mr. Albert Francis    male  26.0      1      1   
8     Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female  18.0      0      0   
9                       Davies, Mr. John Samuel    male  21.0      2      0   

      Ticket     Fare Cabin Embarked  
0     330911   7.8292   NaN        Q  
1     363272   7.0000   NaN        S  
2     240276   9.6875   NaN        Q  
3     315154   8.6625   NaN        S  
4    3101298  12.2875   NaN        S  
5       7538   9.2250   NaN        S  
6     330972   7.6292   NaN        Q  
7     248738  29.0000   NaN        S  
8       2657   7.2292   NaN        C  
9  A/4 48871  24.1500   NaN        S

Выведем третью, десятую, последние строки в одном датафрейме:

In [ ]:
pd.concat([data.iloc[[3]], data[10:11], data[-1:]]) # Первое исполнение, которое было выполенно исходя из инфы в интернете
data.iloc[[3 - 1, 10 - 1, -1]] # Второй вариант, после того как я увидел аналогичное решение в Лаб 1.1

PassengerId  Survived  Pclass                       Name   Sex   Age  \
2            894         0       2  Myles, Mr. Thomas Francis  male  62.0   
9            901         0       3    Davies, Mr. John Samuel  male  21.0   
417         1309         0       3   Peter, Master. Michael J  male   NaN   

     SibSp  Parch     Ticket     Fare Cabin Embarked  
2        0      0     240276   9.6875   NaN        Q  
9        2      0  A/4 48871  24.1500   NaN        S  
417      1      1       2668  22.3583   NaN        C

### **6.Выведите информацию о данных и проанализируйте её. Выведите размер датафрейма. Выведите тип данных каждого столбца. Определите количество пропусков в каждом столбце.**

Информация о данных:

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    object 
 4   Sex          418 non-null    object 
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    object 
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     object 
 11  Embarked     418 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


По ней видно, что в датафрейме 418 строки, проиндексированных от 0 до 417. Можно прочитать названия колонок и тип данных в каждом столбце. Здесь 5 столбцов имеют тип int 64, 5 столбцов типа object, 2 столбца типа float64. Пропузенные значения имаются в столбцах Age, Fare, Cabin.

Размер датафрейма:

In [ ]:
data.shape

(418, 12)

Тип данных каждого столбца:

In [ ]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Количество пропущенных значений в каждом столбце:

In [ ]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

### **7. Определите количество уникальных значений в каждом столбце. Выведите статистики для числовых и не числовых столбцов.**

Количество уникальных значений в каждом столбце:

In [ ]:
data.nunique()

PassengerId    418
Survived         2
Pclass           3
Name           418
Sex              2
Age             79
SibSp            7
Parch            8
Ticket         363
Fare           169
Cabin           76
Embarked         3
dtype: int64

Статистика по числовым столбцам:

In [ ]:
data.describe(include = 'number')

PassengerId    Survived      Pclass         Age       SibSp  \
count   418.000000  418.000000  418.000000  332.000000  418.000000   
mean   1100.500000    0.363636    2.265550   30.272590    0.447368   
std     120.810458    0.481622    0.841838   14.181209    0.896760   
min     892.000000    0.000000    1.000000    0.170000    0.000000   
25%     996.250000    0.000000    1.000000   21.000000    0.000000   
50%    1100.500000    0.000000    3.000000   27.000000    0.000000   
75%    1204.750000    1.000000    3.000000   39.000000    1.000000   
max    1309.000000    1.000000    3.000000   76.000000    8.000000   

            Parch        Fare  
count  418.000000  417.000000  
mean     0.392344   35.627188  
std      0.981429   55.907576  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   31.500000  
max      9.000000  512.329200

Статистика по нечисловым столбцам:

In [ ]:
data.describe(exclude = 'number')
data.describe(include = 'object')

Name   Sex    Ticket            Cabin Embarked
count                        418   418       418               91      418
unique                       418     2       363               76        3
top     Peter, Master. Michael J  male  PC 17608  B57 B59 B63 B66        S
freq                           1   266         5                3      270

### **8. Удалите столбцы с уникальными значениями, а также столбец, где пропущенных значений больше половины.**

Выведем повторно количество уникальных значений по каждому столбцу:

In [ ]:
data.nunique()

PassengerId    418
Survived         2
Pclass           3
Name           418
Sex              2
Age             79
SibSp            7
Parch            8
Ticket         363
Fare           169
Cabin           76
Embarked         3
dtype: int64

Увидим, что столбцы с уникальными значениями это passengerID и Name. Удалим их:

In [ ]:
data = data.drop('PassengerId', axis = 1)
data = data.drop('Name', axis = 1)
data.head(2)

Survived  Pclass     Sex   Age  SibSp  Parch  Ticket    Fare Cabin Embarked
0         0       3    male  34.5      0      0  330911  7.8292   NaN        Q
1         1       3  female  47.0      1      0  363272  7.0000   NaN        S

выведем количество пропущенных значений по столбцам:

In [ ]:
data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age          86
SibSp         0
Parch         0
Ticket        0
Fare          1
Cabin       327
Embarked      0
dtype: int64

Видим, что столбец, в котором отсутствует половина значений лишь 1: Cabin. Удалим этот столбец:

In [ ]:
data = data.drop('Cabin', axis = 1)

Вариант, где меньше человеческого фактора, автоматически удаляем столбцы, где больше 1/2 пропусков.

In [ ]:
data = data.dropna(axis=1, thresh=len(data)/2)

In [ ]:
data.head(2)

Survived  Pclass     Sex   Age  SibSp  Parch  Ticket    Fare Embarked
0         0       3    male  34.5      0      0  330911  7.8292        Q
1         1       3  female  47.0      1      0  363272  7.0000        S

### **9. Замените пропуски в оставшихся столбцах с отсутствующими значениями медианным значением, в случае числового столбца, либо самым распространенным в категориальном столбце.**

Оставшиеся столбцы с пропусками это Age и Fare, оба столбца с числовым типом данных, следовательно, заменим пропущенные значения медианным:

Выведем медианное значение в столбце Age и Fare:

In [ ]:
top_age = data['Age'].mean()
top_fare = data['Fare'].mean()
print(top_age, top_fare)

30.272590361445783 35.627188489208635


Заменим этими значениями

In [ ]:
data['Age'] = data['Age'].fillna(top_age)
data['Fare'] = data['Fare'].fillna(top_fare)

Убедимся, что не осталось отсутсвующихся значений:

In [ ]:
data.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Ticket      0
Fare        0
Embarked    0
dtype: int64

### **10. Переимененуйте все названия столбцов на русский язык.**

Указание:
dict = 'Survived' : 'Выжившие', 'Pclass' : 'Класс пассажира', 'Sex' : 'Пол', 'Age' : ' Возраст', 'SibSp' : 'Количество братьев/сестер', 'Parch' : 'Количество родственников', 'Ticket' : 'Номер билета', 'Fare' : 'Плата за проезд', 'Embarked' : 'Порт посадки'


Переименовываем колонки в соответсвии со словарем:

In [ ]:
data = data.rename(columns={'Survived' : 'Выжившие',
                            'Pclass' : 'Класс пассажира',
                            'Sex' : 'Пол', 'Age' : 'Возраст',
                            'SibSp' : 'Количество братьев/сестер',
                            'Parch' : 'Количество родственников',
                            'Ticket' : 'Номер билета',
                            'Fare' : 'Плата за проезд',
                            'Embarked' : 'Порт посадки'})
data.head(3)

Выжившие  Класс пассажира     Пол  Возраст  Количество братьев/сестер  \
0         0                3    male     34.5                          0   
1         1                3  female     47.0                          1   
2         0                2    male     62.0                          0   

   Количество родственников Номер билета  Плата за проезд Порт посадки  
0                         0       330911           7.8292            Q  
1                         0       363272           7.0000            S  
2                         0       240276           9.6875            Q

### **11. Сохраните получившийся набор данных в формате csv. Считайте корректно полученный файл в новый датафрейм.**

Сохраним датафрейм в формате csv:

In [ ]:
data.to_csv('new_data.csv')

Считаем корректно файл в новый датафрейм:

In [ ]:
new_data = pd.read_csv('new_data.csv', index_col=0)
new_data

Выжившие  Класс пассажира     Пол   Возраст  Количество братьев/сестер  \
0           0                3    male  34.50000                          0   
1           1                3  female  47.00000                          1   
2           0                2    male  62.00000                          0   
3           0                3    male  27.00000                          0   
4           1                3  female  22.00000                          1   
..        ...              ...     ...       ...                        ...   
413         0                3    male  30.27259                          0   
414         1                1  female  39.00000                          0   
415         0                3    male  38.50000                          0   
416         0                3    male  30.27259                          0   
417         0                3    male  30.27259                          1   

     Количество родственников        Номер билета  Плата за проезд  \
0                           0              330911           7.8292   
1                           0              363272           7.0000   
2                           0              240276           9.6875   
3                           0              315154           8.6625   
4                           1             3101298          12.2875   
..                        ...                 ...              ...   
413                         0           A.5. 3236           8.0500   
414                         0            PC 17758         108.9000   
415                         0  SOTON/O.Q. 3101262           7.2500   
416                         0              359309           8.0500   
417                         1                2668          22.3583   

    Порт посадки  
0              Q  
1              S  
2              Q  
3              S  
4              S  
..           ...  
413            S  
414            C  
415            S  
416            S  
417            C  

[418 rows x 9 columns]

### **12. Определите количество выживших. Определить количество выживших моложе 10 лет. Определить количество выживших мужчин, севших в порту Q.**

Выведем строки с выжившими:

In [ ]:
new_data[new_data['Выжившие'] == 1]

Выжившие  Класс пассажира     Пол   Возраст  Количество братьев/сестер  \
1           1                3  female  47.00000                          1   
4           1                3  female  22.00000                          1   
6           1                3  female  30.00000                          0   
8           1                3  female  18.00000                          0   
12          1                1  female  23.00000                          1   
..        ...              ...     ...       ...                        ...   
409         1                3  female   3.00000                          1   
410         1                3  female  30.27259                          0   
411         1                1  female  37.00000                          1   
412         1                3  female  28.00000                          0   
414         1                1  female  39.00000                          0   

     Количество родственников        Номер билета  Плата за проезд  \
1                           0              363272           7.0000   
4                           1             3101298          12.2875   
6                           0              330972           7.6292   
8                           0                2657           7.2292   
12                          0               21228          82.2667   
..                        ...                 ...              ...   
409                         1  SOTON/O.Q. 3101315          13.7750   
410                         0              365237           7.7500   
411                         0               19928          90.0000   
412                         0              347086           7.7750   
414                         0            PC 17758         108.9000   

    Порт посадки  
1              S  
4              S  
6              Q  
8              C  
12             S  
..           ...  
409            S  
410            Q  
411            Q  
412            S  
414            C  

[152 rows x 9 columns]

Их количество:

In [ ]:
new_data[new_data['Выжившие'] == 1].shape[0]

152

Определим количество выживших, младше 10 лет:

In [ ]:
new_data[(new_data['Выжившие'] == 1) & (new_data['Возраст'] < 10)].shape[0]

9

Определим количество выживших мужчин, севших в пороту Q:

In [ ]:
new_data[
    (new_data['Выжившие'] == 1) &
    (new_data['Пол'] == 'male') &
    (new_data['Порт посадки'] == 'Q')
        ].shape[0]

0